# Chapter 6: The Mechanics of Generalization

## 1. Logic Definition of Generalization:

In [1]:
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import LeaveOneOut

def get_min_points_required(X, y):
    """
    Calculates the minimum number of points required to achieve a perfect classification using K-Nearest Neighbors.

    Parameters:
    X (array-like): The input feature matrix.
    y (array-like): The target labels.

    Returns:
    int: The minimum number of points required for perfect classification.
    """
    
    def leave_one_out(X, y):
        if X.shape[0] <= 1:
            return X.shape[0]
        
        loo = LeaveOneOut()

        for train_index, _ in loo.split(X):
            X_train, y_train = X[train_index], y[train_index]
            knn = KNeighborsClassifier(n_neighbors=1)
            knn.fit(X_train, y_train)
            if knn.score(X, y) == 1:
                return leave_one_out(X_train, y_train)
        return X.shape[0]

    return leave_one_out(X, y)

### (a) Show empirically that the information limit of 2 prediction bits per parameter also holds for nearest neighbors.

In [5]:
d_list = [2, 4, 6]
expt_size = 20
class_size = 2

for d in d_list:
    n_full = 2 ** d
    n_avg_list = []
    for _ in range(expt_size):
        X = np.array([list(map(int, np.binary_repr(i, width=d))) for i in range(n_full)])
        y = np.random.randint(class_size, size=n_full)
        n_avg_list.append(get_min_points_required(X, y))
    n_avg = np.mean(n_avg_list)
    print(f"d={d}: n_full={n_full}, Avg. req. points for memorization n_avg={n_avg:.2f}, n_full/n_avg={n_full/n_avg}")

d=2: n_full=4, Avg. req. points for memorization n_avg=2.90, n_full/n_avg=1.3793103448275863
d=4: n_full=16, Avg. req. points for memorization n_avg=8.60, n_full/n_avg=1.8604651162790697
d=6: n_full=64, Avg. req. points for memorization n_avg=32.05, n_full/n_avg=1.9968798751950079


#### Ans: As shown in the printouts above, when d increases, prediction bits per parameter reaches the limit of 2.

### (b) Extend your experiments to multi-class classification.

In [7]:
d_list = [2, 4, 6]
expt_size = 20
class_size = 3

for d in d_list:
    n_full = 2 ** d
    n_avg_list = []
    for _ in range(expt_size):
        X = np.array([list(map(int, np.binary_repr(i, width=d))) for i in range(n_full)])
        y = np.random.randint(class_size, size=n_full)
        n_avg_list.append(get_min_points_required(X, y))
    n_avg = np.mean(n_avg_list)
    print(f"d={d}: n_full={n_full}, Avg. req. points for memorization n_avg={n_avg:.2f}, n_full/n_avg={n_full/n_avg}")

d=2: n_full=4, Avg. req. points for memorization n_avg=3.10, n_full/n_avg=1.2903225806451613
d=4: n_full=16, Avg. req. points for memorization n_avg=11.85, n_full/n_avg=1.350210970464135
d=6: n_full=64, Avg. req. points for memorization n_avg=43.10, n_full/n_avg=1.4849187935034802


#### Ans: As shown in the printouts above for the three-class classification, when d increases, prediction bits per parameter reaches the limit of $\frac{3-1}{2}=1.5$.